forem/forem
# Add (updated) mention autocomplete to post comments

## What type of PR is this? (check all applicable)


## Description

This PR adds the `@mention` autocomplete component to:

- Comments on a post
- Replies to comments on a post
- Replies to post comments in a notification

## Related Tickets & Documents

[RFC: Create Mention Autocomplete](https://github.com/forem/rfcs/pull/16)

## QA Instructions, Screenshots, Recordings

This will need tested in each of the situations mentioned above, ideally on mobile devices as well as desktop to make sure all's working well.

Steps to test post comments:

- Go to a post
- Click to add a comment
- Start typing, including some `@mentions` (you might want to add some users with similar usernames to your local DB to be able to see fuller lists appearing in the dropdown)
- Experiment with selecting options with keyboard (up/down arrows + Enter) and mouse
- Experiment with deleting portions of the mention, ignoring the options and keeping typing, etc - try to break it 😄 
- Repeat as above with replies to comments on a post

Video of these test steps:

https://user-images.githubusercontent.com/20773163/111762231-af286a00-8898-11eb-91ed-1d68a2bfa6fe.mp4


Steps to test post comment replies in notifications:

- Send a notification to your current user in the console:
```
user = User.find_by(username: "admin_mcadmin")
other_user = User.first
article = user.articles.last
comment = article.comments.create!(user: other_user, body_markdown: "Test something")
Notification.send_new_comment_notifications_without_delay(comment)
```
- Refresh the page and go to the notification
- Click reply and repeat the above steps

Video of these test steps:


https://user-images.githubusercontent.com/20773163/111762257-b9e2ff00-8898-11eb-81fd-3510b52d3dc2.mp4


### UI accessibility concerns?

This component should be operable by keyboard alone, and when using a screen reader, should announce the "combobox" functionality, and read the options in the dropdown when populated. This has largely been tested, and concerns documented in the previous PR #13044

## Added tests?


## [Forem core team only] How will this change be communicated?

      or in a [forem.dev](http://forem.dev) post


## [optional] What gif best describes this PR or how it makes you feel?

![David from Schitt's Creeks says yeah you've mentioned that a few times](https://i.giphy.com/media/H1AUCvgHnpRQartFKB/giphy.webp)
